### PR TITLE
Assignment5+6 @Tony Chan

### PR DESCRIPTION
**Pull request by tnadalini**

### CODE REVIEW

**Comment by sammyboy2000** [comment]
I will review this.

**Comment by sammyboy2000** [comment]
1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ✔

   - Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
   - Using LINQ, skip the first row in the `People.csv`. ✔
   - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable (and it may not be depending on how you implement it). ✔

2. Implement `IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()` to return a **sorted**, **unique** list of states. ✔

   - Use `ISampleData.CsvRows` for your data source. ✔
   - Don't forget the list should be unique. ✔
   - Sort the list alphabetically. ✔
   - Include a test that leverages a hardcoded list of Spokane-based addresses. ❌
   - Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ❌

3. Implement `ISampleData.GetAggregateSortedListOfStatesUsingCsvRows()` to return a `string` that contains a **unique**, comma separated list of states. ✔

   - Use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows()` for your data source. ✔
   - Consider "selecting" only the states and calling `ToArray()` to retrieve an array of all the state names. ✔
   - Given the array, consider using `string.Join` to combine the list into a single string. ✔

4. Implement the `ISampleData.People` property to return all the items in `People.csv` as `Person` objects ✔

   - Use `ISampleData.CsvRows` as the source of the data. ✔
   - Sort the list by State, City, Zip. (Sort the addresses first then select). ✔
   - Be sure that `Person.Address` is also populated. ✔
   - Adding null validation to all the `Person` and `Address` properties is **optional**.
   - Consider using `ISampleData.CsvRows` in your test to verify your results. ❌

5. Implement `ISampleDate.FilterByEmailAddress(Predicate<string> filter)` to return a list of names where the email address matches the `filter`. ✔

   - Use `ISampleData.People` for your data source. ✔

6. Implement `ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)` to return a `string` that contains a **unique**, comma separated list of states. ✔

   - Use the `people` parameter from `ISampleData.People` property for your data source. ✔
   - At a minimum, use `System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
   - Don't forget the list should be unique. ✔
   - It is recommended that, at a minimum, you use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` to validate your result. ❌

7. Given the implementation of `Node` in Assignment5

- Implement `IEnumerable<T>` to return all the items in the "circle" of items. ✔
- Add an `IEnumberable<T> ChildItems(int maximum)` method to `Node` that returns the remaining items with a maximum number of items returned less than `maximum`.  ✔

## Extra Credit

- Implement the homework using async/await and multi-threading by defining a new `SampleDataAsync` class that implements `IAsyncSampleData`). Refactor your `SampleData` and `SampleDataAsync` classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌

## Fundamentals

- Ensure you enable:
  - nullable reference types is enabled ✔
  - net6 targeted ✔
  - C# 10.0 ✔
  - and enabled .NET analyzers for both projects ✔
- For this assignment, favor using Assert.AreEqual<T>() (the generic version) ✔
- All of the above should be unit tested ❌
- Choose simplicity over complexity ✔

While the code seems good, more unit tests are needed to meet the requirements. 
Unit tests are needed for the new Node methods.

**Comment by nwnel-017** [comment]
I will be reviewing this one as well, since all other pull requests have been reviewed already. Looks like you need to implement the hard coded unit test using spokane addresses to test part 2. You also need to verify that part 2 is sorted correctly. Besides these issues, your project looks very good to me

File: Assignment/Assignment/SampleData.cs [hunk 1]
```diff
@@ -1,30 +1,101 @@
...
+        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+        {
+            IEnumerable<string> stateList = 
+                (
+                    from line in CsvRows
+                    let state = line.Split(',')
+                    orderby state[6]
+                    select state[6]
+                )
+                .Distinct().ToList();
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:42]:
```suggestion
          return CsvRows
                    .Select(item => item.Split(',')[6])
                    .Distinct()
                    .OrderBy(item => item);
```
It's a bit unexpected to use the query flavor of the LINQ syntax - but if you do I would not recommend mixing the two flavors.

Above is the way to go about it with the method syntax, which is a bit more C# looking and, in my opinion, fits better in your code in terms of readability.

Additionally, you do not need to convert this to a list, as the method returns an `IEnumerable`

File: Assignment/Assignment/SampleData.cs [hunk 2]
```diff
@@ -1,30 +1,101 @@
...
+
+            return stateList;
+        }
 
         // 3.
         public string GetAggregateSortedListOfStatesUsingCsvRows()
-            => throw new NotImplementedException();
+        {
+            string[] states = GetUniqueSortedListOfStatesGivenCsvRows().Select(item => item).ToArray();
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:50]:
```suggestion
            string[] states = GetUniqueSortedListOfStatesGivenCsvRows().ToArray();
```
You don't need this select here.

File: Assignment/Assignment/SampleData.cs [hunk 3]
```diff
@@ -1,30 +1,101 @@
...
+        public IEnumerable<IPerson> People
+        {
+            get
+            {
+                IEnumerable<IPerson> people =
+                    from line in CsvRows
+                    let peopleInfo = line.Split(",")
+                    orderby peopleInfo[5], 
+                            peopleInfo[6], 
+                            peopleInfo[7]
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:65]: This should be ordered by state, city, zip. I believe this is being ordered as city, state, zip.

File: Assignment/Assignment/SampleData.cs [hunk 4]
```diff
@@ -1,30 +1,101 @@
...
+            return filteredList;
+        }
 
         // 6.
-        public string GetAggregateListOfStatesGivenPeopleCollection(
-            IEnumerable<IPerson> people) => throw new NotImplementedException();
+        public string GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)
+        {
+            IEnumerable<string> states = people.Select(item => item.Address.State).Distinct();
+            return states.Aggregate((state1, state2) => state1 + "," + state2);
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:98]: You could combine this into one statement.

**Pull request closed without merging**